File: Queen/lib/queen-code.py
#!/usr/bin/env pythong
"""Queen Code — g16-aware code viewer. No telemetry. Loopback + path jail only."""
from __future__ import annotations

import importlib.util
import json
import os
import re
import subprocess
import sys
from pathlib import Path
from typing import Any

QUEEN = Path(__file__).resolve().parents[1]
SG = QUEEN.parent.parent
_SG_PATHS_LIB = Path(__file__).resolve().parents[2] / "lib"
if str(_SG_PATHS_LIB) not in sys.path:
    sys.path.insert(0, str(_SG_PATHS_LIB))
from sg_paths import grok16_root

GROK16 = grok16_root()
LANG_DOC = QUEEN / "data" / "queen-code-languages.json"
RECENT_DOC = QUEEN / "data" / "queen-code-recent.json"
MAX_READ = int(os.environ.get("QUEEN_CODE_MAX_BYTES", str(2 * 1024 * 1024)))
MAX_RECENT = int(os.environ.get("QUEEN_CODE_MAX_RECENT", "24"))


def _load(path: Path, default: Any = None) -> Any:
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return default if default is not None else {}


def _file_browser():
    spec = importlib.util.spec_from_file_location("qfb", QUEEN / "lib" / "queen-file-browser.py")
    if not spec or not spec.loader:
        return None
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


def _grok16_version() -> dict[str, Any]:
    doc = _load(GROK16 / "data" / "grok16-version.json", {})
    return doc if doc else {"g16_version": "16.2.0", "distro_version": "2.0.0", "discern": []}


def _load_recent() -> list[dict[str, Any]]:
    doc = _load(RECENT_DOC, {"schema": "queen-code-recent/v1", "files": []})
    return list(doc.get("files") or [])


def _save_recent(files: list[dict[str, Any]]) -> None:
    RECENT_DOC.parent.mkdir(parents=True, exist_ok=True)
    RECENT_DOC.write_text(
        json.dumps({"schema": "queen-code-recent/v1", "files": files[:MAX_RECENT]}, indent=2) + "\n",
        encoding="utf-8",
    )


def touch_recent(path: str, *, language: str = "") -> dict[str, Any]:
    real = str(path or "").strip()
    if not real:
        return {"ok": False, "error": "path_required"}
    files = [f for f in _load_recent() if f.get("path") != real]
    files.insert(0, {"path": real, "language": language or _g16_discern(real), "ts": _now_iso()})
    _save_recent(files)
    return {"ok": True, "recent": files[:MAX_RECENT]}


def _now_iso() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def recent_files() -> dict[str, Any]:
    return {"ok": True, "schema": "queen-code-recent/v1", "files": _load_recent()}


def language_registry() -> dict[str, Any]:
    doc = _load(LANG_DOC, {})
    g16 = _grok16_version()
    discern = list(g16.get("discern") or doc.get("g16_discern") or [])
    return {
        "schema": "queen-code/v1",
        "ok": True,
        "title": "Queen Code",
        "motto": doc.get("motto") or "g16 languages · no telemetry",
        "telemetry": False,
        "g16": {
            "version": g16.get("g16_version"),
            "distro": g16.get("distro_version"),
            "discern": discern,
            "belt_profile": (g16.get("belt") or {}).get("default_profile", "belt_2_0"),
        },
        "extensions": doc.get("extensions") or {},
        "profiles": doc.get("profiles") or {},
        "languages": sorted(set(discern + list({v for v in (doc.get("extensions") or {}).values()}))),
    }


def _ext_lang(path: str) -> str:
    doc = _load(LANG_DOC, {})
    ext_map = doc.get("extensions") or {}
    suf = Path(path).suffix
    if suf in ext_map:
        return str(ext_map[suf])
    low = suf.lower()
    for k, v in ext_map.items():
        if k.lower() == low:
            return str(v)
    return "plaintext"


def _g16_discern(path: str) -> str:
    g16 = GROK16 / "bin" / "g16"
    if not g16.is_file():
        return _ext_lang(path)
    try:
        proc = subprocess.run(
            [str(g16), "--g16-discern", path],
            capture_output=True,
            text=True,
            timeout=8,
            check=False,
        )
        if proc.returncode == 0 and proc.stdout.strip():
            return proc.stdout.strip()
    except (OSError, subprocess.TimeoutExpired):
        pass
    return _ext_lang(path)


def _read_text(path: str) -> dict[str, Any]:
    fb = _file_browser()
    if not fb:
        return {"ok": False, "error": "file_browser_missing"}
    resolved = fb.dispatch({"action": "resolve", "path": path})
    if not resolved.get("ok"):
        return resolved
    real = Path(str(resolved.get("resolved") or ""))
    if not real.is_file():
        return {"ok": False, "error": "not_a_file", "path": str(real)}
    size = real.stat().st_size
    if size > MAX_READ:
        return {"ok": False, "error": "file_too_large", "bytes": size, "max": MAX_READ}
    scan = fb.dispatch({"action": "scan", "path": str(real)})
    virus_advisory = None
    if scan and not scan.get("ok"):
        virus_advisory = scan.get("field_virus")
    try:
        text = real.read_text(encoding="utf-8", errors="replace")
    except OSError as exc:
        return {"ok": False, "error": str(exc)}
    lang = _g16_discern(str(real))
    lines = text.count("\n") + (1 if text and not text.endswith("\n") else 0)
    touch_recent(str(real), language=lang)
    return {
        "ok": True,
        "path": str(real),
        "language": lang,
        "g16_discern": lang,
        "bytes": size,
        "lines": lines,
        "content": text,
        "readonly": False,
        "telemetry": False,
        "field_virus_advisory": virus_advisory,
        "recent": _load_recent(),
    }


def _write_text(path: str, content: str) -> dict[str, Any]:
    fb = _file_browser()
    if not fb:
        return {"ok": False, "error": "file_browser_missing"}
    resolved = fb.dispatch({"action": "resolve", "path": path})
    if not resolved.get("ok"):
        return resolved
    real = Path(str(resolved.get("resolved") or ""))
    if len(content.encode("utf-8")) > MAX_READ:
        return {"ok": False, "error": "content_too_large"}
    scan = fb.dispatch({"action": "scan", "path": str(real), "direction": "egress"})
    if scan and not scan.get("ok"):
        return {"ok": False, "error": "field_virus_hold", "field_virus": scan.get("field_virus")}
    try:
        real.parent.mkdir(parents=True, exist_ok=True)
        real.write_text(content, encoding="utf-8")
    except OSError as exc:
        return {"ok": False, "error": str(exc)}
    return {"ok": True, "path": str(real), "bytes": len(content.encode("utf-8")), "saved": True}


def _g16_check(path: str, *, profile: str = "belt_2_0") -> dict[str, Any]:
    g16 = GROK16 / "bin" / "g16"
    if not g16.is_file():
        return {"ok": False, "error": "g16_missing", "hint": "GROK16_ROOT bootstrap"}
    resolved = _read_text(path)
    if not resolved.get("ok"):
        return resolved
    real = Path(resolved["path"])
    lang = resolved.get("g16_discern") or "cxx"
    flags_py = GROK16 / "scripts" / "grok16-profile-flags.py"
    extra = ""
    if flags_py.is_file() and lang in ("c", "cxx"):
        kind = "cxx" if lang == "cxx" else "c"
        try:
            proc = subprocess.run(
                [sys.executable, str(flags_py), profile, kind],
                capture_output=True,
                text=True,
                timeout=15,
                check=False,
                env={**os.environ, "GROK16_ROOT": str(GROK16), "G16_PREFIX": str(GROK16)},
            )
            extra = (proc.stdout or "").strip()
        except (OSError, subprocess.TimeoutExpired):
            pass
    if lang == "python":
        cmd = [str(g16), "-m", "py_compile", str(real)]
    elif lang in ("c", "cxx"):
        out = real.with_suffix(real.suffix + ".queen-code.o")
        cmd = [str(g16), *extra.split(), "-fsyntax-only", str(real)] if extra else [str(g16), "-fsyntax-only", str(real)]
        _ = out
    else:
        return {
            "ok": True,
            "checked": "discern_only",
            "language": lang,
            "profile": profile,
            "message": f"g16 discern={lang} — syntax check via g16 for c/cxx/python",
        }
    try:
        proc = subprocess.run(cmd, capture_output=True, text=True, timeout=60, check=False)
    except (OSError, subprocess.TimeoutExpired) as exc:
        return {"ok": False, "error": str(exc)}
    return {
        "ok": proc.returncode == 0,
        "language": lang,
        "profile": profile,
        "returncode": proc.returncode,
        "stdout": (proc.stdout or "")[-2000:],
        "stderr": (proc.stderr or "")[-4000:],
    }


def _g16_build(path: str, *, profile: str = "belt_2_0") -> dict[str, Any]:
    g16 = GROK16 / "bin" / "g16"
    if not g16.is_file():
        return {"ok": False, "error": "g16_missing"}
    resolved = _read_text(path)
    if not resolved.get("ok"):
        return resolved
    real = Path(resolved["path"])
    lang = resolved.get("g16_discern") or "cxx"
    out_bin = real.with_suffix(real.suffix + ".queen-build")
    flags: list[str] = []
    flags_py = GROK16 / "scripts" / "grok16-profile-flags.py"
    if flags_py.is_file() and lang in ("c", "cxx"):
        kind = "cxx" if lang == "cxx" else "c"
        try:
            proc = subprocess.run(
                [sys.executable, str(flags_py), profile, kind],
                capture_output=True, text=True, timeout=15, check=False,
                env={**os.environ, "GROK16_ROOT": str(GROK16), "G16_PREFIX": str(GROK16)},
            )
            flags = [f for f in (proc.stdout or "").split() if f]
        except (OSError, subprocess.TimeoutExpired):
            pass
    if lang in ("c", "cxx"):
        cmd = [str(g16), *flags, "-o", str(out_bin), str(real)]
    elif lang == "python":
        cmd = [str(g16), "-m", "py_compile", str(real)]
        out_bin = real
    else:
        return {"ok": False, "error": "unsupported_build_lang", "language": lang}
    try:
        proc = subprocess.run(cmd, capture_output=True, text=True, timeout=180, check=False, cwd=str(real.parent))
    except (OSError, subprocess.TimeoutExpired) as exc:
        return {"ok": False, "error": str(exc)}
    return {
        "ok": proc.returncode == 0,
        "language": lang,
        "profile": profile,
        "artifact": str(out_bin),
        "message": f"g16 build {'OK' if proc.returncode == 0 else 'failed'} · {lang} · {profile}",
        "stderr": (proc.stderr or "")[-4000:],
        "stdout": (proc.stdout or "")[-2000:],
    }


def _g16_run_python(path: str, *, profile: str = "belt_2_0") -> dict[str, Any]:
    resolved = _read_text(path)
    if not resolved.get("ok"):
        return resolved
    real = Path(resolved["path"])
    drivers = [
        GROK16 / "bin" / "gpy-16",
        Path(os.environ.get("NEXUS_PYTHONG", "")),
        GROK16 / "python" / "bin" / "pythong",
    ]
    driver = next((d for d in drivers if d and Path(str(d)).is_file()), None)
    if not driver:
        driver = Path(sys.executable)
    env = {**os.environ, "GROK16_ROOT": str(GROK16), "GPY16_FIELD": "1", "PYTHONG_FIELD": "1"}
    try:
        proc = subprocess.run(
            [str(driver), str(real)],
            capture_output=True, text=True, timeout=30, check=False, env=env,
        )
    except (OSError, subprocess.TimeoutExpired) as exc:
        return {"ok": False, "error": str(exc)}
    return {
        "ok": proc.returncode == 0,
        "driver": str(driver),
        "profile": profile,
        "stdout": (proc.stdout or "")[-4000:],
        "stderr": (proc.stderr or "")[-4000:],
    }


def dispatch(body: dict[str, Any]) -> dict[str, Any]:
    action = str(body.get("action") or "status").strip().lower().replace("-", "_")
    if action in ("status", "json"):
        out = language_registry()
        out["recent"] = _load_recent()
        return out
    if action in ("languages", "registry"):
        return language_registry()
    if action in ("recent", "recent_files"):
        return recent_files()
    if action == "touch_recent":
        return touch_recent(str(body.get("path") or ""), language=str(body.get("language") or ""))
    if action in ("read", "open", "load"):
        return _read_text(str(body.get("path") or ""))
    if action in ("write", "save"):
        return _write_text(str(body.get("path") or ""), str(body.get("content") or ""))
    if action in ("discern", "detect_language"):
        path = str(body.get("path") or body.get("filename") or "")
        return {"ok": True, "path": path, "language": _g16_discern(path), "g16_discern": _g16_discern(path)}
    if action in ("g16_check", "check", "compile_check", "syntax"):
        return _g16_check(
            str(body.get("path") or ""),
            profile=str(body.get("profile") or "belt_2_0"),
        )
    if action in ("g16_build", "build"):
        return _g16_build(str(body.get("path") or ""), profile=str(body.get("profile") or "belt_2_0"))
    if action in ("g16_run_python", "run_python"):
        return _g16_run_python(str(body.get("path") or ""), profile=str(body.get("profile") or "belt_2_0"))
    return {
        "ok": False,
        "error": "unknown_action",
        "actions": ["status", "languages", "read", "write", "discern", "g16_check", "g16_build", "g16_run_python", "recent", "touch_recent"],
    }


def main() -> int:
    cmd = (sys.argv[1] if len(sys.argv) > 1 else "json").strip().lower()
    if cmd == "json":
        print(json.dumps(language_registry(), ensure_ascii=False))
        return 0
    if cmd == "dispatch":
        raw = sys.stdin.read()
        body = json.loads(raw) if raw.strip() else {}
        print(json.dumps(dispatch(body), ensure_ascii=False))
        return 0
    print(json.dumps({"error": "usage: queen-code.py [json|dispatch]"}, ensure_ascii=False))
    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: queen/queen-code.js
/**
 * Queen Code — g16-aware viewer. No telemetry. Queen browser native.
 */
(function () {
  "use strict";

  const API = "/api/queen-code";
  const FILES_API = "/api/queen-file-browser";

  const SETTINGS_KEY = "queen-code-settings-v1";
  const RECENT_KEY = "queen-code-recent-v1";

  const defaultSettings = () => ({
    fontSize: 13,
    tabSize: 4,
    wordWrap: false,
    autodetect: true,
    trackRecent: true,
    profile: "belt_2_0",
    theme: "queen_emerald",
    compiler: "belt_2_0",
  });

  const state = {
    path: "",
    language: "plaintext",
    dirty: false,
    langs: [],
    g16: {},
    roots: [],
    settings: defaultSettings(),
    recent: [],
  };

  function $(id) {
    return document.getElementById(id);
  }

  async function codeApi(body) {
    const r = await fetch(API, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    return r.json();
  }

  async function filesApi(body) {
    const r = await fetch(FILES_API, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    return r.json();
  }

  function syncScroll() {
    const ed = $("qc-editor");
    const hi = $("qc-highlight");
    const gut = $("qc-gutter");
    if (!ed || !hi || !gut) return;
    hi.scrollTop = ed.scrollTop;
    hi.scrollLeft = ed.scrollLeft;
    gut.scrollTop = ed.scrollTop;
  }

  function renderGutter(text) {
    const gut = $("qc-gutter");
    if (!gut) return;
    const n = Math.max(1, (text.match(/\n/g) || []).length + (text.length && !text.endsWith("\n") ? 1 : 0));
    const lines = [];
    for (let i = 1; i <= n; i += 1) lines.push(String(i));
    gut.textContent = lines.join("\n");
  }

  function loadSettings() {
    try {
      const raw = localStorage.getItem(SETTINGS_KEY);
      if (raw) state.settings = { ...defaultSettings(), ...JSON.parse(raw) };
    } catch (_) {
      state.settings = defaultSettings();
    }
    applySettings();
  }

  function saveSettings() {
    localStorage.setItem(SETTINGS_KEY, JSON.stringify(state.settings));
    applySettings();
  }

  function applySettings() {
    const s = state.settings;
    const wrap = document.querySelector(".qc-editor-wrap");
    const ed = $("qc-editor");
    const hi = $("qc-highlight");
    const gut = $("qc-gutter");
    const px = `${s.fontSize}px`;
    if (wrap) wrap.classList.toggle("wrap", !!s.wordWrap);
    [ed, hi, gut].forEach((el) => {
      if (!el) return;
      el.style.fontSize = px;
      el.style.tabSize = String(s.tabSize);
    });
    $("qc-set-font") && ($("qc-set-font").value = String(s.fontSize));
    $("qc-set-tab") && ($("qc-set-tab").value = String(s.tabSize));
    $("qc-set-wrap") && ($("qc-set-wrap").checked = !!s.wordWrap);
    $("qc-set-autodetect") && ($("qc-set-autodetect").checked = !!s.autodetect);
    $("qc-set-recent") && ($("qc-set-recent").checked = !!s.trackRecent);
    $("qc-set-profile") && ($("qc-set-profile").value = s.profile || s.compiler || "belt_2_0");
    $("qc-set-theme") && ($("qc-set-theme").value = s.theme || "queen_emerald");
    $("qc-menu-compiler") && ($("qc-menu-compiler").value = s.compiler || s.profile || "belt_2_0");
    document.body.dataset.qcTheme = s.theme || "queen_emerald";
  }

  function loadRecentLocal() {
    try {
      const raw = localStorage.getItem(RECENT_KEY);
      if (raw) state.recent = JSON.parse(raw) || [];
    } catch (_) {
      state.recent = [];
    }
  }

  function saveRecentLocal() {
    localStorage.setItem(RECENT_KEY, JSON.stringify(state.recent.slice(0, 24)));
    renderRecentMenu();
  }

  function pushRecent(path, language) {
    if (!state.settings.trackRecent || !path) return;
    state.recent = [{ path, language: language || state.language, ts: Date.now() },
      ...state.recent.filter((r) => r.path !== path)].slice(0, 24);
    saveRecentLocal();
  }

  function renderRecentMenu() {
    const el = $("qc-recent-list");
    if (!el) return;
    const items = state.recent || [];
    el.innerHTML = items.length
      ? items.map((r) =>
          `<button type="button" data-recent="${encodeURIComponent(r.path)}">${r.path.split("/").pop()}</button>`,
        ).join("")
      : `<span class="qc-menu-label">No recent files</span>`;
    el.querySelectorAll("[data-recent]").forEach((btn) => {
      btn.addEventListener("click", () => openFile(decodeURIComponent(btn.dataset.recent || "")));
    });
  }

  async function autodetectLanguage(path) {
    if (!state.settings.autodetect || !path) return state.language;
    const doc = await codeApi({ action: "discern", path });
    return doc.language || doc.g16_discern || state.language;
  }

  function paint() {
    const ed = $("qc-editor");
    const hi = $("qc-highlight");
    if (!ed || !hi) return;
    const text = ed.value;
    renderGutter(text);
    hi.innerHTML = window.QueenCodeHighlight
      ? window.QueenCodeHighlight.highlight(text, state.language)
      : text.replace(/&/g, "&amp;").replace(/</g, "&lt;");
    syncScroll();
    updateStatus();
  }

  function updateStatus() {
    const ed = $("qc-editor");
    const text = ed?.value || "";
    const lines = text.split("\n").length;
    const cols = (text.split("\n").pop() || "").length + 1;
    $("qc-st-lang").textContent = state.language;
    $("qc-st-g16").textContent = state.g16.version || "g16";
    $("qc-st-pos").textContent = `Ln ${lines}, Col ${cols}`;
    $("qc-st-path").textContent = state.path ? state.path.split("/").pop() : "untitled";
    $("qc-st-dirty").textContent = state.dirty ? "● modified" : "";
    $("qc-tab-label").textContent = state.path ? state.path.split("/").pop() : "Welcome";
  }

  function toast(msg, ok) {
    const el = $("qc-toast");
    if (!el) return;
    el.textContent = msg;
    el.style.borderColor = ok ? "#238636" : "#f85149";
    el.hidden = false;
    clearTimeout(toast._t);
    toast._t = setTimeout(() => {
      el.hidden = true;
    }, 4000);
  }

  async function loadRegistry() {
    const doc = await codeApi({ action: "languages" });
    state.langs = doc.languages || [];
    state.g16 = doc.g16 || {};
    if (window.QueenCodeHighlight?.mergeExtensions && doc.extensions) {
      window.QueenCodeHighlight.mergeExtensions(doc.extensions);
    }
    const list = $("qc-lang-list");
    if (list) {
      list.innerHTML = (state.langs || [])
        .map((l) => `<span title="g16 discern">${l}</span>`)
        .join(" · ");
    }
    $("qc-g16-pill").textContent = `g16 ${doc.g16?.version || ""} · belt ${doc.g16?.belt_profile || "2.0"}`;
    $("qc-telemetry-pill").classList.add("ok");
    $("qc-telemetry-pill").textContent = "telemetry: off";
  }

  async function loadRoots() {
    const doc = await filesApi({ action: "roots" });
    state.roots = doc.roots || [];
    const tree = $("qc-tree");
    if (!tree) return;
    tree.innerHTML = "";
    for (const r of state.roots) {
      const btn = document.createElement("button");
      btn.type = "button";
      btn.textContent = `📁 ${r.label}`;
      btn.addEventListener("click", () => browseRoot(r.path));
      tree.appendChild(btn);
    }
  }

  async function browseRoot(path) {
    const doc = await filesApi({ action: "list", path });
    const tree = $("qc-tree");
    if (!tree || !doc.ok) return;
    tree.innerHTML = "";
    const back = document.createElement("button");
    back.type = "button";
    back.textContent = "← roots";
    back.addEventListener("click", loadRoots);
    tree.appendChild(back);
    for (const e of doc.entries || []) {
      const btn = document.createElement("button");
      btn.type = "button";
      btn.textContent = `${e.kind === "dir" ? "📁" : "📄"} ${e.name}`;
      btn.addEventListener("click", () => {
        if (e.kind === "dir") browseRoot(e.path);
        else openFile(e.path);
      });
      tree.appendChild(btn);
    }
  }

  async function openFile(path) {
    const doc = await codeApi({ action: "read", path });
    if (!doc.ok) {
      toast(doc.error || "read failed", false);
      return;
    }
    state.path = doc.path;
    state.language = doc.language || doc.g16_discern || "plaintext";
    state.dirty = false;
    $("qc-editor").value = doc.content || "";
    if (doc.recent) state.recent = doc.recent;
    pushRecent(state.path, state.language);
    paint();
    toast(`Opened · ${state.language} · g16 discern`, true);
  }

  async function saveFile() {
    if (!state.path) {
      toast("No path — open a file from explorer", false);
      return;
    }
    const content = $("qc-editor").value;
    const doc = await codeApi({ action: "write", path: state.path, content });
    if (!doc.ok) {
      toast(doc.error || "save failed", false);
      return;
    }
    state.dirty = false;
    updateStatus();
    toast("Saved", true);
  }

  function activeProfile() {
    return state.settings.compiler || state.settings.profile || "belt_2_0";
  }

  async function g16Check() {
    if (!state.path) {
      toast("Open a file first", false);
      return;
    }
    await saveFile();
    const doc = await codeApi({ action: "g16_check", path: state.path, profile: activeProfile() });
    if (doc.ok) toast(`g16 check OK · ${doc.language}`, true);
    else toast((doc.stderr || doc.error || "g16 check failed").slice(0, 200), false);
  }

  async function g16Build() {
    if (!state.path) {
      toast("Open a source file first", false);
      return;
    }
    await saveFile();
    toast("Building…", true);
    const doc = await codeApi({ action: "g16_build", path: state.path, profile: activeProfile() });
    if (doc.ok) toast(doc.message || `Build OK · ${doc.artifact || ""}`, true);
    else toast((doc.stderr || doc.error || "build failed").slice(0, 240), false);
  }

  async function g16Launch() {
    const launchPath = state.path?.endsWith(".launch") ? state.path : state.path?.replace(/\.[^.]+$/, ".launch");
    if (!launchPath) {
      toast("Open a .launch chamber or source with matching .launch", false);
      return;
    }
    toast("Launching chamber…", true);
    const doc = await filesApi({ action: "run_launch", path: launchPath, profile: activeProfile() });
    if (doc.ok) toast(doc.message || "Launch OK", true);
    else toast((doc.error || doc.stderr || "launch failed").slice(0, 240), false);
  }

  async function runPython() {
    if (!state.path) {
      toast("Open a .py file first", false);
      return;
    }
    await saveFile();
    const doc = await codeApi({ action: "g16_run_python", path: state.path, profile: activeProfile() });
    if (doc.ok) toast(`Python OK · ${(doc.stdout || "").slice(0, 120)}`, true);
    else toast((doc.stderr || doc.error || "run failed").slice(0, 240), false);
  }

  function bindMenus() {
    $("qc-menu-save")?.addEventListener("click", saveFile);
    $("qc-menu-g16")?.addEventListener("click", g16Check);
    $("qc-menu-open-files")?.addEventListener("click", () => $("qc-open-files")?.click());
    $("qc-menu-wrap")?.addEventListener("click", () => {
      state.settings.wordWrap = !state.settings.wordWrap;
      saveSettings();
    });
    $("qc-menu-font-inc")?.addEventListener("click", () => {
      state.settings.fontSize = Math.min(22, (state.settings.fontSize || 13) + 1);
      saveSettings();
    });
    $("qc-menu-font-dec")?.addEventListener("click", () => {
      state.settings.fontSize = Math.max(11, (state.settings.fontSize || 13) - 1);
      saveSettings();
    });
    $("qc-set-font")?.addEventListener("input", (e) => {
      state.settings.fontSize = parseInt(e.target.value, 10) || 13;
      saveSettings();
    });
    $("qc-set-tab")?.addEventListener("change", (e) => {
      state.settings.tabSize = parseInt(e.target.value, 10) || 4;
      saveSettings();
    });
    $("qc-set-wrap")?.addEventListener("change", (e) => {
      state.settings.wordWrap = !!e.target.checked;
      saveSettings();
    });
    $("qc-set-autodetect")?.addEventListener("change", (e) => {
      state.settings.autodetect = !!e.target.checked;
      saveSettings();
    });
    $("qc-set-recent")?.addEventListener("change", (e) => {
      state.settings.trackRecent = !!e.target.checked;
      saveSettings();
    });
    $("qc-set-profile")?.addEventListener("change", (e) => {
      state.settings.profile = e.target.value || "belt_2_0";
      state.settings.compiler = state.settings.profile;
      saveSettings();
    });
    $("qc-set-theme")?.addEventListener("change", (e) => {
      state.settings.theme = e.target.value || "queen_emerald";
      saveSettings();
    });
    $("qc-menu-compiler")?.addEventListener("change", (e) => {
      state.settings.compiler = e.target.value || "belt_2_0";
      state.settings.profile = state.settings.compiler;
      saveSettings();
    });
    $("qc-menu-build")?.addEventListener("click", g16Build);
    $("qc-menu-launch")?.addEventListener("click", g16Launch);
    $("qc-menu-run-py")?.addEventListener("click", runPython);
    $("qc-build")?.addEventListener("click", g16Build);
    $("qc-launch")?.addEventListener("click", g16Launch);
  }

  function bindEditor() {
    const ed = $("qc-editor");
    ed?.addEventListener("input", () => {
      state.dirty = true;
      paint();
    });
    ed?.addEventListener("scroll", syncScroll);
    ed?.addEventListener("keydown", (e) => {
      if ((e.ctrlKey || e.metaKey) && e.key === "s") {
        e.preventDefault();
        saveFile();
      }
    });
    $("qc-save")?.addEventListener("click", saveFile);
    $("qc-act-g16")?.addEventListener("click", g16Check);
    document.querySelector(".qc-toolbar #qc-g16")?.addEventListener("click", g16Check);
    $("qc-open-files")?.addEventListener("click", () => {
      const url = `${location.origin}/world/queen-files.html`;
      try {
        if (parent?.QueenOS?.browser?.navigate) parent.QueenOS.browser.navigate(url);
        else location.href = url;
      } catch (_) {
        location.href = url;
      }
    });
    $("qc-open-browser")?.addEventListener("click", () => {
      const url = `${location.origin}/world/browser.html`;
      try {
        if (parent?.QueenOS?.browser?.navigate) parent.QueenOS.browser.navigate(url);
        else location.href = url;
      } catch (_) {
        location.href = url;
      }
    });
  }

  function welcomeText() {
    return [
      "// Queen Code — rewritten viewer for g16",
      "// No telemetry · loopback only · VSCodium-inspired layout",
      "//",
      "// Open a file from the explorer or Queen Files.",
      "// Languages: " + (state.langs || []).slice(0, 8).join(", ") + "…",
      "",
      "/* g16 discern drives syntax + compile check */",
    ].join("\n");
  }

  async function init() {
    loadSettings();
    loadRecentLocal();
    bindMenus();
    bindEditor();
    await loadRegistry();
    const recentDoc = await codeApi({ action: "recent" });
    if (recentDoc.files?.length) state.recent = recentDoc.files;
    renderRecentMenu();
    await loadRoots();
    const params = new URLSearchParams(location.search);
    const path = params.get("path");
    if (path) await openFile(path);
    else {
      $("qc-editor").value = welcomeText();
      state.language = "cxx";
      paint();
    }
  }

  init();
})();